### PR TITLE
(test): Add Functions Router foundry tests

### PR DESCRIPTION
**Pull request by justinkaseman**
```
| File                                                                                   | % Lines           | % Statements      | % Branches      | % Funcs         |
| src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol                                       | 90.91% (100/110)  | 92.06% (116/126)  | 76.47% (26/34)  | 100.00% (22/22) |
```

### CODE REVIEW

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol [hunk 1]
```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
...
+      FunctionsRequest.CodeLanguage.JavaScript,
+      sourceCode
+    );
+    bytes memory requestData = FunctionsRequest.encodeCBOR(request);
+
+    vm.expectRevert(FunctionsSubscriptions.InvalidConsumer.selector);
+    s_functionsRouter.sendRequest(s_subscriptionId, requestData, FunctionsRequest.REQUEST_DATA_VERSION, 5000, s_donId);
+  }
+
+  function test_SendRequest_RevertIfInvalidCallbackGasLimit() public {
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:309]: Can we add a test case for reversions if the subscription balance isn't sufficient?
We could test both the cases where the user sets the callback gas limit to something reasonable like 100k and the case where the gas limit is zero just to confirm expected behavior.

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol [hunk 2]
```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
...
+    Vm.Log[] memory entries = vm.getRecordedLogs();
+    bytes32 requestIdFromEvent = entries[2].topics[1];
+
+    assertEq(requestIdFromReturn, requestIdFromEvent);
+  }
 }
 
 /// @notice #sendRequestToProposed
-contract FunctionsRouter_SendRequestToProposed {
+contract FunctionsRouter_SendRequestToProposed is FunctionsSubscriptionSetup {
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:429]: The additional test cases I mentioned above can be added for sendRequestToProposed too.

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol [hunk 3]
```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
...
+      costWithoutCallback,
+      transmitter,
+      commitment
+    );
+
+    assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.INVALID_COMMITMENT));
+    assertEq(callbackGasCostJuels, 0);
+  }
+
+  function test_Fulfill_RequestNotProcessedInsufficientGas() public {
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:824]: The commented out test cases are pretty important.  What will is the blocker for getting these working?

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol [hunk 4]
```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
...
+    //   response,
+    //   err,
+    //   juelsPerGas,
+    //   costWithoutCallback,
+    //   transmitter,
+    //   commitment
+    // );
+    // assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.COST_EXCEEDS_COMMITMENT));
+    // assertEq(callbackGasCostJuels, 0);
+  }
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:930]: Can we add a test for the consumer's callback reverting?  We could do this by setting a low callback gas limit instead of needing to create a new test consumer contract with a reversion in the callback.

**Review by KuphJr — COMMENTED** [review]
Had a few more suggested test cases.  Otherwise it LGTM

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol [hunk 5]
```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
...
+
+    FunctionsRouter.Config memory config = s_functionsRouter.getConfig();
+    uint32[] memory _maxCallbackGasLimits = config.maxCallbackGasLimits;
+    uint32 maxCallbackGasLimit = _maxCallbackGasLimits[callbackGasLimitsIndexSelector];
 
+    vm.expectRevert(abi.encodeWithSelector(FunctionsRouter.GasLimitTooBig.selector, maxCallbackGasLimit));
+    s_functionsRouter.isValidCallbackGasLimit(s_subscriptionId, maxCallbackGasLimit + 1);
+  }
+
+  function test_IsValidCallbackGasLimit_Success() public view {
```
[the remark this answers is not in the document]
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:125]: Since it runs out of gas it returns a `FulfillResult.USER_CALLBACK_ERROR`
But the `callbackReturnData` because it doesn't even have enough gas to return that.

Makes me think maybe we should have a check in there for a minimum amount of gas so that the user can at least get an error back

[the remark this answers is not in the document]
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:125]: Added a test case so we can observe this: `test_Fulfill_SuccessUserCallbackRunsOutOfGas`

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 1 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:309]: Just added one for gas limit = 0, but for the other one could you explain more what you mean?
Is that not `test_Fulfill_RequestNotProcessedSubscriptionBalanceInvariant`?

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 4 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:930]: Covered by `test_Fulfill_SuccessUserCallbackReverts`

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 2 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:429]: Added

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 3 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:824]: I'll get everything but the `test_Fulfill_RequestNotProcessedSubscriptionBalanceInvariant` one up, because that one requires some helper contracts to be written

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (KuphJr) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:127]: What I meant was just the behavior of `s_functionsRouter.isValidCallbackGasLimit(subId, 0)`.  This could be a potential test case.

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 1 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:311]: "do you mean the same test for insufficient sub balance but with two different callbackGasLimits in the request?"

Yes.  I was just thinking of the potential edge case where the gas limit specified is 0, but the user still doesn't have a large enough subscription balance to cover the baseline cost.
If it makes more sense, we can add the tests in FunctionsCoordinator.t.sol.

File: contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:991]: Ah, I missed that because I was thinking `test_Fulfill_SuccessUserCallbackError` meant a successful callback function call, but with an error message.

Therefore, do we have a test case which covers successfully calling the callback and providing an error message instead of a response?

Comment (justinkaseman) [reply contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol:991]: Yeah that naming wasn't great hah. I'll add that
